src/fenwick.jl

```diff
@@ -0,0 +1,114 @@
+"""
+    FenwickTree{T}(n)
+    
+Constructs a [`FenwickTree`](https://en.wikipedia.org/wiki/Fenwick_tree) of length `n`.
+ 
+"""
+mutable struct FenwickTree{T}
+    bi_tree::Vector{T} #bi_tree is shorthand for Binary Indexed Tree, an alternative name for Fenwick Tree
+    n::Integer
+end
+
+FenwickTree{T}() where T = FenwickTree{T}(0)
+FenwickTree{T}(n::Integer) where T = FenwickTree{T}(zeros(T, n), n)
+
+"""
+    FenwickTree(arr::AbstractArray) 
+    
+Constructs a [`FenwickTree`](https://en.wikipedia.org/wiki/Fenwick_tree) from `arr`.
+ 
+"""
+function FenwickTree(a::AbstractVector{U}) where U
+    n = size(a)[1]
+    tree = FenwickTree{U}(n)
+    @inbounds for i = 1:n
+        inc!(tree, i, a[i])
+    end
+    tree
+end
+
+length(ft::FenwickTree{T}) where T = ft.n
+
+"""
+    inc!(ft::FenwickTree{T}, ind, val)
+
+Increases the value of the [`FenwickTree`] by `val` from the index `ind` upto the length of the Fenwick Tree.
+
+""" 
+function inc!(ft::FenwickTree{T}, ind::Int, val) where T
+    val0 = convert(T, val)
+    i = ind
+    n = ft.n
+    @boundscheck 1 <= i <= n || throw(ArgumentError("$i should be in between 1 and $n"))
+    @inbounds while i <= n
+        ft.bi_tree[i] += val0
+        i += i&(-i)
+    end
+end
+
+"""
+    dec!(ft::FenwickTree{T}, ind, val)
+
+Decreases the value of the [`FenwickTree`] by `val` from the index `ind` upto the length of the Fenwick Tree.
+
+""" 
+dec!(ft::FenwickTree{T}, ind::Int, val) where T = inc!(ft, ind, -val)
+
+"""
+    inc!(ft::FenwickTree{T}, range::AbstractUnitRange, val)
+
+Increases the value of the [`FenwickTree`] by `val` from the indices in `range`.
+
+"""    
+function inc!(ft::FenwickTree{T}, range::AbstractUnitRange, val) where T
```

Comment:
I suggested this change, but I'm wondering if updating a range of values in a Fenwick tree is a common operation?

Looking at its use in the tests, my intuition says that this should be equivalent to
```
for idx in range
    inc!(ft, idx, val)
end
```
which is clearly not the case.

Given this is actually equivalent to an `inc!` followed by a `dec!`, how about renaming this function `incdec!` (and not using a range).  I.e.,

```suggestion
function incdec!(ft::FenwickTree{T}, left::Integer, right::Integer, val)
```

The intuition here is that the value at `left` is being increased by `val`, which increases the prefix sum for everything, but then the value at `right` is decreased by `val`, leaving the remaining elements unchanged.

If you follow (and agree with) this intuition, then that actually changes the implementation below slightly (i.e., the value at `right` would be decreased / decremented instead of `right +1`).

(I'm open to other ideas...)

Reply:
Umm, the range update above as you've mentioned(`for idx in range`) is an `O((r - l + 1) * log n)` operation, whereas what I implemented here is `O(logn)` which is much better. The intuition is that of a [Heaviside step function](https://en.wikipedia.org/wiki/Heaviside_step_function), you add `val` from `left` to `Inf`, and subtract `val` from `right + 1` to `Inf`, what remains after this is `+val` from `left` to `right`.

As for `incdec!`, this is just opposite of what I stated above.

Read [this](https://ocw.mit.edu/courses/mathematics/18-03sc-differential-equations-fall-2011/unit-iii-fourier-series-and-laplace-transform/step-and-delta-functions-integrals-and-generalized-derivatives/MIT18_03SCF11_s24_1text.pdf). The exact thing I'm referring to is the `Box function`.

Reply:
I think I'm using a different mental model of Fenwick trees than you, so please bear with me.  

From reading about Fenwick trees, it seems that their purpose is to keep counts of numbers in such a way that it is easy to get the cumulative sum (or prefix sum) (though not necessarily the original count).  So if I'm keeping track counts of 5 numbers, I might have 

```julia
x = [1, 0, 0, 3, 2]   # 1 one, 3 fours, and 2 fives
cumsum(x) # = [1, 1, 1, 4, 6]
```

I can then increment the third element:
```julia
x[3] += 1  # x = [1, 0, 1, 3, 2]
cumsum(x) # = [1, 1, 2, 5, 7]
```

So what I see a Fenwick tree doing is exactly the same, except that it calculates/updates individual values of the cumulative sum more efficiently (and doesn't keep track of the actual counts).

```julia
ft = FenwickTree([1, 0, 0, 3, 2])
[ft[i] for i = 1:5]  # [1, 1, 1, 4, 6]
inc!(ft, 3, 1);
[ft[i] for i = 1:5]  # [1, 1, 2, 5, 7]
```

This is why `inc!` and `dec!` make sense to me (they update the underlying counts), and why I proposed `incdec!` as a way to updating the cumulative sum (or prefix sum) for a range in the middle of the array.

From your response, it seems that in your mental model, you're trying to manipulate the cumulative sums directly.  Can you explain why viewing the data structure in this way is useful?

Reply:
First of all, I'm definitely in for adding the functionality of `incdec!`, but what I'm suggesting is just a operational benefit, nothing more nothing less. If range updates are not there as per my model, complexity of the operation, i.e update each index individually is an expensive operation, additionally what I'm suggesting is making a sort of `arr[l : r] += val` , i know this thing doesn't exist in Julia, but there's no harm in introducing it. Actually the use case becomes more evident, when we have range queries, but for that we need another manipulation. I suggest we scrap this functionality, 'cuz we are introducing too much complications here. Instead, we add the `incdec!`

Reply:
I think my main concern with the current implementation is this:

```julia
ft = FenwickTree([1, 0, 0, 3, 2])
inc!(ft, 3, 1);
[ft[i] for i = 1:5]  # [1, 1, 2, 5, 7]
```
vs.

```julia
ft = FenwickTree([1, 0, 0, 3, 2])
inc!(ft, 3:5, 1)
[ft[i] for i = 1:5]  # [1, 1, 2, 5, 7]
```

These two operations do the same thing with the current code, but I don't think they should.

(BTW, adding a value to a range of a Vector is easy in Julia (though you need to use `.+`:

```julia
julia> x = [1, 0, 0, 3, 2]
5-element Array{Int64,1}:
 1
 0
 0
 3
 2

julia> x[2:4] .+= 1; x
5-element Array{Int64,1}:
 1
 1
 1
 4
 2
```
)

Reply:
Umm, I can't seem to explain the behavior right now. Like, I've spent more than 1 hr trying to understand why it's happening, but now I've understood. The output will be same, 'cuz you're doing the same operation. Please read the docstring. 
```
ft = FenwickTree([1, 0, 0, 3, 2])
inc!(ft, 3:4, 1);
[ft[i] for i = 1:5] #[1, 1, 2, 5, 6]
```

Reply:
Right.  I do understand why they're doing the same thing in the current code.  But I don't think they should do the same thing. :-)

For consistency,

```julia
inc!(ft, 3:4, 1);
```
should be equivalent to
```julia
inc!(ft, 3, 1);
inc!(ft, 4, 1);
```
And it obviously isn't.  But, it is equivalent to 
```julia
inc!(ft, 3, 1);
dec!(ft, 5, 1);
```
Or, if you were to add `incdec!`, 
```julia
incdec!(ft, 3, 5, 1)
```